### PR TITLE
net/tstun: use gaissmai/bart instead of tempfork/device

### PR DESCRIPTION
**Pull request by oxtoacart**
This implementation uses less memory than tempfork/device, which helps avoid OOM conditions in the iOS VPN extension when switching to a Tailnet with ExitNode routing enabled.

Updates tailscale/corp#18514

### CODE REVIEW

File: net/tstun/table/table.go [hunk 1]
```diff
@@ -73,7 +58,7 @@ func (t *RoutingTableBuilder) Build() *RoutingTable {
 //
 // It is safe for concurrent use.
 type RoutingTable struct {
-	prefixTrie *device.AllowedIPs
+	table *bart.Table[key.NodePublic]
```
Comment (maisem) [inline net/tstun/table/table.go:61]: the caller really just wants the addresses but we had to go through the `key.NodePublic` abstraction previously. See my other PR, we should really just have `bart.Table[netip.Addr]` here

Comment (maisem) [reply net/tstun/table/table.go:61]: or in the layer above, we could just delete this package

Comment (oxtoacart) [reply net/tstun/table/table.go:61]: The thing that this abstraction adds is the login in `InsertOrReplace` that handles changed prefixes for a given peer by first deleting the old ones before adding the new ones. I noticed that your PR dropped that and I wasn't sure that was safe to do, as I don't know what exactly depends on that behavior but I assume it was done for a reason.

Comment (maisem) [reply net/tstun/table/table.go:61]: The whole table is atomically replaced in tstun. We don't use the replace logic at all

Comment (oxtoacart) [reply net/tstun/table/table.go:61]: Coo, so `wcfg.Peers` is guaranteed not to contain duplicates?

Comment (maisem) [reply net/tstun/table/table.go:61]: Yep

Comment (oxtoacart) [reply net/tstun/table/table.go]: Done

File: net/tstun/wrap.go [hunk 2]
```diff
@@ -619,7 +619,7 @@ type natFamilyConfig struct {
 	// dstAddrToPeerKeyMapper is the routing table used to map a given dst IP to
 	// the peer key responsible for that IP.
 	// It only contains peers that require a MasqueradeAsIP address.
-	dstAddrToPeerKeyMapper *table.RoutingTable
+	dstAddrToPeerKeyMapper *bart.Table[key.NodePublic]
```
Comment (maisem) [inline net/tstun/wrap.go:622]: I think you can combine `dstMasqAddrs` and `dstAddrToPeerKeyMapper` and just have a single `*bart.Table[netip.Addr]`

Comment (oxtoacart) [reply net/tstun/wrap.go]: Done, thanks for your patch!

File: net/tstun/wrap.go [hunk 3]
```diff
@@ -640,21 +635,21 @@ func (c *natFamilyConfig) String() string {
 		i++
 		return true
 	})
-	count := map[netip.Addr]int{}
-	c.dstMasqAddrs.Range(func(_ key.NodePublic, v netip.Addr) bool {
-		count[v]++
-		return true
-	})
+	// count := map[netip.Addr]int{}
```
Comment (maisem) [inline net/tstun/wrap.go:638]: Let's not have commented out code here. This was to get the build passing in the patch 😅
We can track the counts when building the table and then just use the value directly here

File: net/tstun/wrap.go [hunk 4]
```diff
@@ -747,17 +739,18 @@ func natConfigFromWGConfig(wcfg *wgcfg.Config, addrFam ipproto.Version) *natFami
...
 		}
-		rt.InsertOrReplace(p.PublicKey, p.AllowedIPs...)
-		mak.Set(&dstMasqAddrs, p.PublicKey, addrToUse)
+		for _, ip := range p.AllowedIPs {
+			needMasq = true
+			rt.Insert(ip, addrToUse)
+		}
 	}
-	if len(listenAddrs) == 0 && len(dstMasqAddrs) == 0 {
+	if len(listenAddrs) == 0 && needMasq {
```
Comment (maisem) [inline net/tstun/wrap.go:747]: !needMasq

File: net/tstun/wrap.go [hunk 5]
```diff
@@ -747,17 +739,18 @@ func natConfigFromWGConfig(wcfg *wgcfg.Config, addrFam ipproto.Version) *natFami
 		} else {
 			continue
 		}
-		rt.InsertOrReplace(p.PublicKey, p.AllowedIPs...)
-		mak.Set(&dstMasqAddrs, p.PublicKey, addrToUse)
+		for _, ip := range p.AllowedIPs {
+			needMasq = true
```
Comment (maisem) [inline net/tstun/wrap.go:743]: Can probably do this outside the allowed ips loops

**Comment by oxtoacart** [comment]
@maisem I've addressed your latest comments.

**Review by maisem — APPROVED** [review]
Thanks!

**Comment by oxtoacart** [comment]
Thanks for the review @maisem !

**Pull request merged**